modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientPrimaryReplicaTracker.java

```diff
@@ -0,0 +1,342 @@
+    void stop() {
+        catalogService.removeListener(CatalogEvent.TABLE_DROP, (EventListener) this);
+        placementDriver.removeListener(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, (EventListener) this);
+        primaryReplicas.clear();
+    }
+
+    @Override
+    public CompletableFuture<Boolean> notify(EventParameters parameters, @Nullable Throwable exception) {
+        if (exception != null) {
+            return CompletableFuture.completedFuture(false);
+        }
+
+        if (parameters instanceof DropTableEventParameters) {
+            removeTable((DropTableEventParameters) parameters);
+
+            return CompletableFuture.completedFuture(false);
+        }
+
+        if (!(parameters instanceof PrimaryReplicaEventParameters)) {
+            assert false : "Unexpected event parameters: " + parameters.getClass();
+
+            return CompletableFuture.completedFuture(false);
+        }
+
+        PrimaryReplicaEventParameters primaryReplicaEvent = (PrimaryReplicaEventParameters) parameters;
+        if (!(primaryReplicaEvent.groupId() instanceof TablePartitionId)) {
+            return CompletableFuture.completedFuture(false);
+        }
+
+        TablePartitionId tablePartitionId = (TablePartitionId) primaryReplicaEvent.groupId();
+        updatePrimaryReplica(tablePartitionId, primaryReplicaEvent.startTime(), primaryReplicaEvent.leaseholder());
+
+        return CompletableFuture.completedFuture(false); // false: don't remove listener.
+    }
+
+    private void removeTable(DropTableEventParameters dropTableEvent) {
+        // Use previous version of the catalog to get the dropped table.
+        int prevCatalogVersion = dropTableEvent.catalogVersion() - 1;
+
+        CatalogTableDescriptor table = catalogService.table(dropTableEvent.tableId(), prevCatalogVersion);
+        assert table != null : "Table from DropTableEventParameters not found: " + dropTableEvent.tableId();
+
+        CatalogZoneDescriptor zone = catalogService.zone(table.zoneId(), prevCatalogVersion);
+        assert zone != null : "Zone from DropTableEventParameters not found: " + table.zoneId();
+
+        for (int partition = 0; partition < zone.partitions(); partition++) {
+            TablePartitionId tablePartitionId = new TablePartitionId(dropTableEvent.tableId(), partition);
+            primaryReplicas.remove(tablePartitionId);
+        }
+    }
+
+    private void updatePrimaryReplica(TablePartitionId tablePartitionId, HybridTimestamp startTime, String nodeName) {
+        long startTimeLong = startTime.longValue();
+
+        primaryReplicas.compute(tablePartitionId, (key, existingVal) -> {
+            if (existingVal != null
+                    && existingVal.leaseStartTime != null
+                    && existingVal.leaseStartTime.longValue() >= startTimeLong) {
+                return existingVal;
+            }
+
+            return new ReplicaHolder(nodeName, startTime);
+        });
+
+        while (true) {
```

Comment:
Here and there. Seems that it's possible to use `maxStartTime.updateAndGet(value -> Math.max(value, startTimeLong));` instead.

Reply:
Nice! Did not know about that. Fixed.

Should we (later) update other places like `HybridClockImpl`?

Reply:
We either should, or maybe @denis-chudov can tell us that while(true) is faster (which is important in HybridClock and not important here).